docs/notebooks/asynchronous_bayesian_optimization.pct.py

```diff
@@ -0,0 +1,373 @@
+
+# %%
+from trieste.space import Box
+from trieste.data import Dataset
+from trieste.objectives import SCALED_BRANIN_MINIMUM
+
+search_space = Box([0, 0], [1, 1])
+num_initial_points = 3
+initial_query_points = search_space.sample(num_initial_points)
+initial_observations = objective(initial_query_points.numpy(), sleep=False)
+initial_data = Dataset(
+    query_points=initial_query_points,
+    observations=tf.constant(initial_observations, dtype=tf.float64),
+)
+
+import gpflow
+from trieste.models import create_model
+from trieste.models.gpflow.config import GPflowModelConfig
+
+
+def build_model(data):
+    variance = tf.math.reduce_variance(data.observations)
+    kernel = gpflow.kernels.RBF(variance=variance)
+    gpr = gpflow.models.GPR(data.astuple(), kernel, noise_variance=1e-5)
+    gpflow.set_trainable(gpr.likelihood, False)
+
+    return GPflowModelConfig(
+        **{
+            "model": gpr,
+            "optimizer": gpflow.optimizers.Scipy(),
+            "optimizer_args": {
+                "minimize_args": {"options": dict(maxiter=100)},
+            },
+        }
+    )
+
+# these imports will be used later for optimization
+from trieste.acquisition import LocalPenalizationAcquisitionFunction
+from trieste.acquisition.rule import AsyncEfficientGlobalOptimization, EfficientGlobalOptimization
+from trieste.ask_tell_optimization import AskTellOptimizer
+
+
+# %% [markdown]
+# ## Multiprocessing setup
+#
+# To keep this notebook as reproducible as possible, we will only be using Python's multiprocessing package here. In this section we will explain our setup and define some common code to be used later.
+#
+# In both synchronous and asynchronous scenarios we will have a fixed set of worker processes performing observations. We will also have a main process responsible for optimization process with Trieste. When Trieste suggests a new point, it is inserted into a points queue. One of the workers picks this point from the queue, performs the observation, and inserts the output into observations queue. The main process then picks up the observation from the queue, at which moment it either waits for the rest of the points in the batch to come back (synchronous scenario) or immediately suggests a new point (asynchrunous scenario). This process continues either for a certain number of iterations or until we accumulate necessary number of observations.
+# 
+# The overall setup is illustrated in this diagram:
+# ![multiprocessing setup](figures/async_bo.png)
+
+# %%
+# Necessary multiprocessing primitives
+from multiprocessing import Manager, Process
+
+# %% [markdown]
+# We now define several common functions to implement the described setup. First we define a worker function that will be running a single observation in a separate process. Worker takes both queues as an input, reads next point from the points queue, makes an observation, and inserts observed data into the observations queue.
```

Comment:
not sure if its needed, but perhaps it would be useful to add something about assumptions - that the user actually has multiple threads/cores

perhaps add a link to a useful python documentation on multiprocessing

Reply:
No assumptions really, that code can run anywhere where Python can. But in case of 1 CPU there will be no performance gain, in fact it may degrade. I'll add a short note in the last section.